### PR TITLE
[DeviceMSAN] Support MemorySanitizer for device offloading

### PR DESCRIPTION
**Pull request by AllanZyne**
LLVM: https://github.com/intel/llvm/pull/15955

### CODE REVIEW

File: source/loader/layers/sanitizer/asan/asan_ddi.cpp
Comment (zhaomaosu) [inline source/loader/layers/sanitizer/asan/asan_ddi.cpp:1873]: Should it be Asan? BTW 'initMsanDDITable' is not user friendly.

File: source/loader/layers/sanitizer/msan/msan_buffer.cpp
Comment (zhaomaosu) [inline source/loader/layers/sanitizer/msan/msan_buffer.cpp:9]:
```suggestion
 * @file msan_buffer.cpp
```

File: source/loader/layers/sanitizer/msan/msan_ddi.cpp
Comment (zhaomaosu) [inline source/loader/layers/sanitizer/msan/msan_ddi.cpp:78]: Added a fixme for these code?

File: source/loader/layers/sanitizer/asan/asan_ddi.cpp
Comment (yingcong-wu) [inline source/loader/layers/sanitizer/asan/asan_ddi.cpp:1873]: Why it logs "Msan" in asan_ddi.cpp?

File: source/loader/layers/sanitizer/msan/msan_buffer.cpp
Comment (yingcong-wu) [inline source/loader/layers/sanitizer/msan/msan_buffer.cpp:9]:
```suggestion
 * @file msan_buffer.cpp
```

File: source/loader/layers/sanitizer/msan/msan_interceptor.cpp
Comment (zhaomaosu) [inline source/loader/layers/sanitizer/msan/msan_interceptor.cpp:4]:
```suggestion
 * Copyright (C) 2024 Intel Corporation
```

File: source/loader/layers/sanitizer/msan/msan_interceptor.cpp
Comment (zhaomaosu) [inline source/loader/layers/sanitizer/msan/msan_interceptor.cpp:47]: memory sanitizer doesn't use redone, looks like the comments can be removed

File: source/loader/layers/sanitizer/msan/msan_interceptor.cpp
Comment (zhaomaosu) [inline source/loader/layers/sanitizer/msan/msan_interceptor.cpp:159]: remove loop

File: source/loader/layers/sanitizer/msan/msan_interceptor.cpp
Comment (zhaomaosu) [inline source/loader/layers/sanitizer/msan/msan_interceptor.cpp:453]: remove unused code?

File: source/loader/layers/sanitizer/msan/msan_ddi.cpp
Comment (yingcong-wu) [inline source/loader/layers/sanitizer/msan/msan_ddi.cpp:79]: Could we add some comments here to help other understand why we need these keys?

File: source/loader/layers/sanitizer/msan/msan_interceptor.hpp
Comment (zhaomaosu) [inline source/loader/layers/sanitizer/msan/msan_interceptor.hpp:3]:
```suggestion
 * Copyright (C) 2024 Intel Corporation
```

File: source/loader/layers/sanitizer/msan/msan_interceptor.hpp
Comment (zhaomaosu) [inline source/loader/layers/sanitizer/msan/msan_interceptor.hpp:82]: remove unused code

File: source/loader/layers/sanitizer/msan/msan_interceptor.hpp
Comment (zhaomaosu) [inline source/loader/layers/sanitizer/msan/msan_interceptor.hpp:173]: looks like this struct can be removed

File: source/loader/layers/sanitizer/sanitizer_common/sanitizer_shadow.hpp
Comment (zhaomaosu) [inline source/loader/layers/sanitizer/sanitizer_common/sanitizer_shadow.hpp:9]: Why we need this header file?

File: source/loader/layers/sanitizer/msan/msan_interceptor.hpp
Comment (yingcong-wu) [inline source/loader/layers/sanitizer/msan/msan_interceptor.hpp:3]:
```suggestion
 * Copyright (C) 2024 Intel Corporation
```

File: source/loader/layers/sanitizer/msan/msan_report.cpp
Comment (yingcong-wu) [inline source/loader/layers/sanitizer/msan/msan_report.cpp:33]: If we use Memory sanitizer here, should we update the DevASAN? To speak for myself, I think we can use "DeviceSanitizer" for all sanitizers.

File: source/loader/layers/sanitizer/msan/msan_shadow.hpp
Comment (yingcong-wu) [inline source/loader/layers/sanitizer/msan/msan_shadow.hpp:53]: Do you mean "CPU/PVC" or CPU Device?

File: source/loader/layers/sanitizer/msan/msan_shadow.hpp
Comment (yingcong-wu) [inline source/loader/layers/sanitizer/msan/msan_shadow.hpp:138]: Not matching the code.

File: source/loader/layers/sanitizer/sanitizer_common/sanitizer_shadow.hpp
Comment (yingcong-wu) [inline source/loader/layers/sanitizer/sanitizer_common/sanitizer_shadow.hpp:9]:
```suggestion
 * @file sanitizer_shadow.hpp
```

**Review by yingcong-wu — APPROVED** [review]

**Review by zhaomaosu — APPROVED** [review]
lgtm

**Comment by AllanZyne** [comment]
Hi @oneapi-src/unified-runtime-maintain, internal review has done. Please review, thank you.

**Comment by AllanZyne** [comment]
CI "Build and test / E2E OpenCL / Start e2e job / Build SYCL, UR, run E2E" failed is due to opencl driver haven't upgraded to 2025.0.

**Comment by cdai2** [comment]
ping @oneapi-src/unified-runtime-maintain

**Comment by kbenzie** [comment]
> ping @oneapi-src/unified-runtime-maintain

https://github.com/intel/llvm/pull/15955 is *Waiting on code owner review from intel/dpcpp-clang-driver-reviewers, intel/dpcpp-esimd-reviewers, intel/dpcpp-tools-reviewers, intel/llvm-reviewers-runtime* until those code owners approve that PR this won't be merged.

**Comment by AllanZyne** [comment]
Hi @oneapi-src/unified-runtime-maintain, I have got almost all reviewers approve, could you please review this PR as well?
Thank you!

File: source/loader/layers/sanitizer/ur_sanddi.cpp
Comment (pbalcer) [inline source/loader/layers/sanitizer/ur_sanddi.cpp:22]:
```
bool asan_enabled = enabledLayerNames.count("UR_LAYER_MSAN");
bool msan_enabled = enabledLayerNames.count("UR_LAYER_MSAN");

if (asan_enabled && msan_enabled) {
  getContext()->logger.warning("Enabling MSAN and ASAN at the same time is not supported.")
  return UR_RESULT_SUCCESS;
} else if (asan_enabled) {
  ...
}
```

Just a suggestion

File: source/loader/layers/sanitizer/msan/msan_interceptor.cpp
Comment (pbalcer) [inline source/loader/layers/sanitizer/msan/msan_interceptor.cpp:166]: This, and a few other functions in this file, are nearly straight copy-paste from asan interceptor.

I know there's time pressure now, but please create a high-priority ticket to refactor this post merge. Maybe there should be a common base class for sanitizers that contains the shared code?

File: source/loader/layers/sanitizer/msan/msan_interceptor.cpp
Comment (pbalcer) [inline source/loader/layers/sanitizer/msan/msan_interceptor.cpp:351]: please remove this loop.
I also noticed that the same thing is done in the same asan function. Please remove that as well.

File: source/loader/layers/sanitizer/msan/msan_options.cpp
Comment (pbalcer) [inline source/loader/layers/sanitizer/msan/msan_options.cpp:25]: This is a near 1:1 copy/paste from asan_options.

File: source/loader/layers/sanitizer/msan/msan_shadow.cpp
Comment (pbalcer) [inline source/loader/layers/sanitizer/msan/msan_shadow.cpp:109]: create a define for the magic number

**Review by pbalcer — COMMENTED** [review]
Mostly lgtm. My biggest concern is the large amounts of copy/pasting that the new msan is doing from the existing asan codebase. I understand the urgency for now, so we can move go with the current implementation, but please consider how you could reduce the number of duplications in your codebase this is creating so that asan/msan are easier to maintain long-term.

File: source/loader/layers/sanitizer/msan/msan_interceptor.cpp
Comment (AllanZyne) [reply source/loader/layers/sanitizer/msan/msan_interceptor.cpp:166]: Sure, we noticed this problem as well, after intense period of feature developing, we will try to merge common codes to make it more easily maintain.

File: source/loader/layers/sanitizer/msan/msan_options.cpp
Comment (AllanZyne) [reply source/loader/layers/sanitizer/msan/msan_options.cpp:25]: We consider to use a single sanitizer option in the future

**Review by pbalcer — APPROVED** [review]

**Comment by AllanZyne** [comment]
> Mostly lgtm. My biggest concern is the large amounts of copy/pasting that the new msan is doing from the existing asan codebase. I understand the urgency for now, so we can move with the current implementation, but please consider how you could reduce the number of duplications in your codebase this is creating so that asan/msan are easier to maintain long-term.

Thank you for your understanding, @pbalcer!

**Comment by kbenzie** [comment]
Once the L0 jobs complete I'll squash the commits then merge this.

**Comment by kbenzie** [comment]
After merging https://github.com/oneapi-src/unified-runtime/pull/2420 (since the CI was ready) there is a conflict on `source/loader/layers/sanitizer/asan/asan_ddi.cpp` please fix @AllanZyne and I'll merge this first thing in the morning UK time.

**Comment by AllanZyne** [comment]
> After merging #2420 (since the CI was ready) there is a conflict on `source/loader/layers/sanitizer/asan/asan_ddi.cpp` please fix @AllanZyne and I'll merge this first thing in the morning UK time.

Hi @kbenzie, I have to merge main branch to solve the conflict, which is not the commit id pointed by LLVM.
Hope the CI in LLVM works well.

**Pull request merged**
